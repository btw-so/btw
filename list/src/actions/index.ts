export * from './alerts';
export * from './app';
export * from './github';
export * from './user';
export * from './list';
export * from './notes';