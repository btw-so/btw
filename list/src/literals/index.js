import { keyMirror } from "@gilbarbara/helpers";

// import { Status } from "types";

export const ActionTypes = keyMirror({
  GITHUB_GET_REPOS_REQUEST: undefined,
  GITHUB_GET_REPOS_SUCCESS: undefined,
  GITHUB_GET_REPOS_FAILURE: undefined,
  HIDE_ALERT: undefined,
  SHOW_ALERT: undefined,
  SET_APP_OPTIONS: undefined,
  USER_LOGIN_REQUEST: undefined,
  USER_LOGIN_SUCCESS: undefined,
  USER_LOGIN_FAILURE: undefined,
  USER_LOGOUT_REQUEST: undefined,
  USER_LOGOUT_SUCCESS: undefined,
  USER_LOGOUT_FAILURE: undefined,
  GET_USER: undefined,
  GET_USER_SUCCESS: undefined,
  GET_USER_FAILURE: undefined,
  GENERATE_OTP: undefined,
  GENERATE_OTP_SUCCESS: undefined,
  GENERATE_OTP_FAILURE: undefined,
  VERIFY_OTP: undefined,
  VERIFY_OTP_SUCCESS: undefined,
  VERIFY_OTP_FAILURE: undefined,
  UPDATE_USER: undefined,
  UPDATE_USER_SUCCESS: undefined,
  UPDATE_USER_FAILURE: undefined,
  RESET_USER_STATE: undefined,
  GET_LIST: undefined,
  GET_LIST_SUCCESS: undefined,
  GET_LIST_FAILURE: undefined,
  NEW_LIST_NODE: undefined,
  EDIT_LIST_NODE_POS: undefined,
  EDIT_LIST_NODE_CONTENT: undefined,
  UPSERT_LIST_NODE: undefined,
  CHANGE_SELECTED_NODE: undefined,
  BATCH_PUSH_NODES: undefined,
  SAVE_NOTE_CONTENT: undefined,
  GET_NOTE: undefined,
  GET_NOTE_SUCCESS: undefined,
  GET_NOTE_FAILURE: undefined,
  ADD_FILE: undefined,
  ADD_FILE_SUCCESS: undefined,
  ADD_FILE_FAILURE: undefined,
  GET_FILE: undefined,
  GET_FILE_SUCCESS: undefined,
  GET_FILE_FAILURE: undefined,
  GET_PINNED_NODES: undefined,
  GET_PINNED_NODES_SUCCESS: undefined,
  GET_PINNED_NODES_FAILURE: undefined,
  GET_PUBLIC_NOTE: undefined,
  GET_PUBLIC_NOTE_SUCCESS: undefined,
  GET_PUBLIC_NOTE_FAILURE: undefined,
  SEARCH_NODES: undefined,
  SEARCH_NODES_SUCCESS: undefined,
  SEARCH_NODES_FAILURE: undefined,
});

export const STATUS = {
  IDLE: "idle",
  RUNNING: "running",
  READY: "ready",
  SUCCESS: "success",
  ERROR: "error",
  PINNED: "pinned",
};
