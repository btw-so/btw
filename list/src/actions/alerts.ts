import React from 'react';
import { uuid } from '@gilbarbara/helpers';
import { createAction } from '@reduxjs/toolkit';

import { actionPayload } from 'modules/helpers';

import { ActionTypes } from 'literals';

import { ShowAlertOptions } from 'types';

export const hideAlert = createAction<string>(ActionTypes.HIDE_ALERT);

export const showAlert = createAction(
  ActionTypes.SHOW_ALERT,
  (message: React.ReactNode, options: ShowAlertOptions) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return actionPayload({
      id: options.id || uuid(),
      icon: options.icon || 'dot-circle-o',
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    });
  },
);
