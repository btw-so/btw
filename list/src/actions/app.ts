import { createAction } from '@reduxjs/toolkit';

import { ActionTypes } from 'literals';

import { SetAppOOptions } from 'types';

export const setAppOptions = createAction<SetAppOOptions>(ActionTypes.SET_APP_OPTIONS);
