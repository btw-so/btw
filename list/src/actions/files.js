import { createAction } from "@reduxjs/toolkit";

import { ActionTypes } from "literals";


export const addFile = createAction(
  ActionTypes.ADD_FILE,
  (data) => ({
    payload: data,
  })
);
export const addFileSuccess = createAction(
  ActionTypes.ADD_FILE_SUCCESS,
  (data) => ({
    payload: data,
  })
);
export const addFileFailure = createAction(
  ActionTypes.ADD_FILE_FAILURE,
  (data) => ({
    payload: data,
  })
);

export const getFile = createAction(ActionTypes.GET_FILE, (data) => ({
  payload: data,
}));
export const getFileSuccess = createAction(
  ActionTypes.GET_FILE_SUCCESS,
  (data) => ({
    payload: data,
  })
);
export const getFileFailure = createAction(
  ActionTypes.GET_FILE_FAILURE,
  (data) => ({
    payload: data,
  })
);
