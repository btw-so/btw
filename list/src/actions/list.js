import { createAction } from "@reduxjs/toolkit";

import { ActionTypes } from "literals";

export const batchPushNodes = createAction(
  ActionTypes.BATCH_PUSH_NODES,
  (data) => ({
    payload: data,
  })
);

export const getList = createAction(ActionTypes.GET_LIST, (data) => ({
  payload: data,
}));
export const getListSuccess = createAction(
  ActionTypes.GET_LIST_SUCCESS,
  (data) => ({
    payload: data,
  })
);
export const getListFailure = createAction(
  ActionTypes.GET_LIST_FAILURE,
  (data) => ({
    payload: data,
  })
);

export const newListNode = createAction(ActionTypes.NEW_LIST_NODE, (data) => ({
  payload: data,
}));

export const editListNodePos = createAction(
  ActionTypes.EDIT_LIST_NODE_POS,
  (data) => ({
    payload: data,
  })
);

export const editListNodeContent = createAction(
  ActionTypes.EDIT_LIST_NODE_CONTENT,
  (data) => ({
    payload: data,
  })
);

export const upsertListNode = createAction(
  ActionTypes.UPSERT_LIST_NODE,
  (data) => ({
    payload: data,
  })
);

export const changeSelectedNode = createAction(
  ActionTypes.CHANGE_SELECTED_NODE,
  (data) => ({
    payload: data,
  })
);

export const getPinnedNodes = createAction(
  ActionTypes.GET_PINNED_NODES,
  (data) => ({
    payload: data,
  })
);

export const getPinnedNodesSuccess = createAction(
  ActionTypes.GET_PINNED_NODES_SUCCESS,
  (data) => ({
    payload: data,
  })
);

export const getPinnedNodesFailure = createAction(
  ActionTypes.GET_PINNED_NODES_FAILURE,
  (data) => ({
    payload: data,
  })
);

export const getPublicNote = createAction(
  ActionTypes.GET_PUBLIC_NOTE,
  (data) => ({
    payload: data,
  })
);

export const getPublicNoteSuccess = createAction(
  ActionTypes.GET_PUBLIC_NOTE_SUCCESS,
  (data) => ({
    payload: data,
  })
);

export const getPublicNoteFailure = createAction(
  ActionTypes.GET_PUBLIC_NOTE_FAILURE,
  (data) => ({
    payload: data,
  })
);

export const searchNodes = createAction(
  ActionTypes.SEARCH_NODES,
  (data) => ({
    payload: data,
  })
);

export const searchNodesSuccess = createAction(
  ActionTypes.SEARCH_NODES_SUCCESS,
  (data) => ({
    payload: data,
  })
);

export const searchNodesFailure = createAction(
  ActionTypes.SEARCH_NODES_FAILURE,
  (data) => ({
    payload: data,
  })
);
