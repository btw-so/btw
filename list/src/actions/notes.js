import { createAction } from "@reduxjs/toolkit";

import { ActionTypes } from "literals";


export const saveNoteContent = createAction(
  ActionTypes.SAVE_NOTE_CONTENT,
  (data) => ({
    payload: data,
  })
);

export const getNote = createAction(ActionTypes.GET_NOTE, (data) => ({
  payload: data,
}));
export const getNoteSuccess = createAction(
  ActionTypes.GET_NOTE_SUCCESS,
  (data) => ({
    payload: data,
  })
);
export const getNoteFailure = createAction(
  ActionTypes.GET_NOTE_FAILURE,
  (data) => ({
    payload: data,
  })
);
