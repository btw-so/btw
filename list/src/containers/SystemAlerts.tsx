import React, { useCallback, useEffect, useRef } from 'react';
import { useDispatch } from 'react-redux';
import { useUnmount } from 'react-use';
import styled from 'styled-components';
import { responsive } from 'styled-minimal';

import { useAppSelector } from 'modules/hooks';
import { spacer } from 'modules/theme';

import { hideAlert } from 'actions';

import Alert from 'components/Alert';
import Transition from 'components/Transition';

const Base = styled.div`
  position: fixed;
  z-index: 1000;

  > div {
    > * + * {
      margin-top: ${spacer(3)};
    }
  }
`;

const TopLeft = styled(Base)`
  left: ${spacer(3)};
  top: ${spacer(3)};
  width: 26rem;

  ${
    /* sc-custom '@media-query' */ responsive({
      md: {
        width: '32rem',
      },
    })
  };
`;

const TopRight = styled(Base)`
  right: ${spacer(3)};
  top: ${spacer(3)};
  width: 26rem;

  ${
    /* sc-custom '@media-query' */ responsive({
      md: {
        width: '32rem',
      },
    })
  };
`;

const BottomLeft = styled(Base)`
  bottom: ${spacer(3)};
  left: ${spacer(3)};
  width: 26rem;

  ${
    /* sc-custom '@media-query' */ responsive({
      md: {
        width: '32rem',
      },
    })
  };
`;

const BottomRight = styled(Base)`
  bottom: ${spacer(3)};
  right: ${spacer(3)};
  width: 26rem;

  ${
    /* sc-custom '@media-query' */ responsive({
      md: {
        width: '32rem',
      },
    })
  };
`;

const SystemAlertsWrapper = styled.div`
  pointer-events: none;
  position: fixed;
  z-index: 1000;
`;

export default function SystemAlerts() {
  const dispatch = useDispatch();
  const alerts = useAppSelector(s => s.alerts.data);
  const timeouts: Record<string, any> = useRef({});

  useEffect(() => {
    const { current } = timeouts;

    if (alerts.length) {
      alerts.forEach(d => {
        if (d.timeout && !current[d.id]) {
          current[d.id] = setTimeout(() => {
            dispatch(hideAlert(d.id));
          }, d.timeout * 1000);
        }
      });
    }
  }, [alerts, dispatch]);

  useUnmount(() => {
    const { current } = timeouts;

    Object.keys(current).forEach(d => {
      clearTimeout(current[d]);
    });
  });

  const handleClick = useCallback(
    (event: React.MouseEvent<HTMLButtonElement>) => {
      event.preventDefault();
      const { id = '' } = event.currentTarget.dataset;

      dispatch(hideAlert(id));
    },
    [dispatch],
  );

  const renderAlerts = useCallback(
    (position: string) => {
      const items = alerts.filter(d => d.position === position);

      if (!items.length) {
        return null;
      }

      return items.map(d => (
        <Alert
          key={d.id}
          handleClickClose={handleClick}
          icon={d.icon}
          id={d.id}
          variant={d.variant}
        >
          {d.message}
        </Alert>
      ));
    },
    [alerts, handleClick],
  );

  return (
    <SystemAlertsWrapper key="SystemAlerts">
      <TopLeft>
        <Transition transition="slideRight">{renderAlerts('top-left')}</Transition>
      </TopLeft>
      <TopRight>
        <Transition transition="slideLeft">{renderAlerts('top-right')}</Transition>
      </TopRight>
      <BottomLeft>
        <Transition transition="slideRight">{renderAlerts('bottom-left')}</Transition>
      </BottomLeft>
      <BottomRight>
        <Transition transition="slideLeft">{renderAlerts('bottom-right')}</Transition>
      </BottomRight>
    </SystemAlertsWrapper>
  );
}
