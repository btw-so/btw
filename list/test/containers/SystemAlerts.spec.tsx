import React from 'react';
import { config } from 'react-transition-group';

import { ActionTypes } from 'literals';

import { showAlert } from 'actions';

import SystemAlerts from 'containers/SystemAlerts';

import { fireEvent, render, screen } from 'test-utils';

config.disabled = true;

jest.mock('components/Transition', () => ({ children }) => (
  <div className="transition">{children}</div>
));

jest.useFakeTimers();

const mockDispatch = jest.fn();

describe('SystemAlerts', () => {
  it('should render all zones', () => {
    const { container } = render(<SystemAlerts />);

    expect(container).toMatchSnapshot();
  });

  it('should render an alert and hide itself after the timeout', async () => {
    render(<SystemAlerts />, {
      actions: [
        showAlert('Hello World', {
          position: 'top-left',
          variant: 'success',
        }),
      ],
      mockDispatch,
    });

    expect(screen.getByRole('alert')).toBeInTheDocument();

    jest.runOnlyPendingTimers();

    expect(mockDispatch).toHaveBeenCalledWith({
      type: ActionTypes.HIDE_ALERT,
      payload: '8cdee72f-28d4-4441-91f0-c61f6e3d9684',
    });

    expect(screen.queryByRole('alert')).not.toBeInTheDocument();
  });

  it('should render an alert without timeout and close it', () => {
    render(<SystemAlerts />, {
      actions: [
        showAlert('Hello Mars', {
          id: 'ABD13',
          position: 'bottom-right',
          variant: 'dark',
          timeout: 0,
        }),
      ],
      mockDispatch,
    });

    expect(screen.getByRole('alert')).toBeInTheDocument();

    fireEvent.click(screen.queryByRole('alert')?.querySelector('button')!);

    expect(screen.queryByRole('alert')).not.toBeInTheDocument();
  });
});
