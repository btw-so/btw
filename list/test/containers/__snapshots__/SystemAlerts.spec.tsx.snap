// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SystemAlerts should render all zones 1`] = `
.c1 {
  position: fixed;
  z-index: 1000;
}

.c1 > div > * + * {
  margin-top: 16px;
}

.c2 {
  left: 16px;
  top: 16px;
  width: 26rem;
}

.c3 {
  right: 16px;
  top: 16px;
  width: 26rem;
}

.c4 {
  bottom: 16px;
  left: 16px;
  width: 26rem;
}

.c5 {
  bottom: 16px;
  right: 16px;
  width: 26rem;
}

.c0 {
  pointer-events: none;
  position: fixed;
  z-index: 1000;
}

@media screen and (min-width:768px) {
  .c2 {
    width: 32rem;
  }
}

@media screen and (min-width:768px) {
  .c3 {
    width: 32rem;
  }
}

@media screen and (min-width:768px) {
  .c4 {
    width: 32rem;
  }
}

@media screen and (min-width:768px) {
  .c5 {
    width: 32rem;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <div
        class="transition"
      />
    </div>
    <div
      class="c1 c3"
    >
      <div
        class="transition"
      />
    </div>
    <div
      class="c1 c4"
    >
      <div
        class="transition"
      />
    </div>
    <div
      class="c1 c5"
    >
      <div
        class="transition"
      />
    </div>
  </div>
</div>
`;
